app/Listeners/ValidateSNMP.php

```diff
@@ -0,0 +1,153 @@
+     */
+    public function __construct()
+    {
+        //
+    }
+
+    /**
+     * Handle the event.
+     *
+     * @param  CreatingDevice  $event
+     * @return void
+     */
+    public function handle(CreatingDevice $event)
+    {
+        if (host_exists($event->device->hostname)) {
+            throw new HostExistsException("Already have host {$event->device->hostname}");
+        }
+        
+        if (!is_numeric($event->device->port_association_mode)) {
+            $event->device->port_association_mode = ($event->device->port_association_mode ?: 'ifIndex');
+            
+            try {
+                $event->device->port_association_mode = \App\Models\PortAssociationMode::where('name', $event->device->port_association_mode)->first()->pom_id;
+            } catch (\Exception $e) {
+                throw new InvalidPortAssocModeException("Invalid port_association_mode id '{$event->device->port_association_mode}'. Valid modes are: " . join(', ', get_port_assoc_modes()));
+            }
+        }
+
+        // Valid port assoc mode ID
+        if (!\App\Models\PortAssociationMode::exists($event->device->port_association_mode)) {
+            throw new InvalidPortAssocModeException("Invalid port_association_mode id '{$event->device->port_association_mode}'. Valid modes are: " . join(', ', get_port_assoc_modes()));
+        }
+        // check if we have the host by IP
+        if (Config::get('addhost_alwayscheckip') === true) {
+            $ip = gethostbyname($event->device->hostname);
+        } else {
+            $ip = $event->device->hostname;
+        }
+        if ($event->device->force_add !== true && ip_exists($ip)) {
+            throw new HostIpExistsException("Already have host with this IP {$event->device->hostname}");
+        }
+
+        // Test reachability
+        if (!$event->device->force_add) {
+            $address_family = snmpTransportToAddressFamily($event->device->transport);
+            $ping_result = isPingable($event->device->hostname, $address_family);
+            if (!$ping_result['result']) {
+                throw new HostUnreachablePingException("Could not ping {$event->device->hostname}");
+            }
+        }
+
+        // if $snmpver isn't set, try each version of snmp
+        if (empty($event->device->snmpver)) {
+            $snmpvers = Config::get('snmp.version');
+        } else {
+            $snmpvers = array($event->device->snmpver);
+        }
+
+        $host_unreachable_exception = new HostUnreachableException("Could not connect to {$event->device->hostname}, please check the snmp details and snmp reachability");
+        $valid = false;
+
+        foreach ($snmpvers as $snmpver) {
+            if ($snmpver === "v3") {
+                // Try each set of parameters from config
+                $event->device->snmpver = $snmpver;
+                $v3 = Config::get('snmp.v3');
+                if ($event->device->snmpV3AuthSet()) {
+                    array_push($v3, [
```

Comment:
Should this be pushed to the Config like it does in the UI?

Reply:
It depends on if it is used elsewhere in the code.